feat: Option to set expiry for a shareable link of a document

surajshetty3416 opened the pull request:
- By default, all the new links generated (sent via email) will be expired in 90 days (this is configurable).
- Shareable link generated from a document will have to option to set custom expiry.


https://user-images.githubusercontent.com/13928957/156182902-c02b3bd2-5320-4cda-bae7-98b2f428fdb2.mov



---

**Fixes:**
284079
284081

---

**ToDo:**
- [x] Page to show link expired & invalid key page.
<img width="400" alt="Screenshot 2022-03-01 at 7 33 35 PM" src="https://user-images.githubusercontent.com/13928957/156183468-f0b2fb52-db76-42ee-935b-e772c014a35e.png">
<img width="400" alt="Screenshot 2022-03-01 at 7 33 51 PM" src="https://user-images.githubusercontent.com/13928957/156183605-92b41269-8b4d-46e5-b970-a1bfbfbcb277.png">


(After review from Sagar & Faris got following actionable)
- [x] "Share link" option only for users with share permission
- [x] ~Modal to show all generated links (show links with expiry and creator of the link)~ not required
- [x] Rename to “Document Share Key”
- [x] “Allow Older Web View Links (Insecure)” option in System Settings
- [x] Option to Generate link without expiry.
- [x] Remove `document_key_expiry` from doctype and move it to System Settings
- [x] Print and Download option in print view (optional)
- [x] Unique constraint for the generated key (optional)
- [x] Use existing key if expiry is same (optional)

surajshetty3416 committed:
fix: Maintain document signature in Document Key table

surajshetty3416 committed:
fix: Allow old signature as well

surajshetty3416 committed:
feat: Document key expiry

surajshetty3416 committed:
feat: Set expiry to document link

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into fix-document-signature

surajshetty3416 committed:
fix: Logic to show link expired message

surajshetty3416 committed:
style: Fix sider warnings

netchampfaris commented on frappe/core/doctype/document_key/document_key.json:
```diff
@@ -0,0 +1,74 @@
+ "modified": "2022-01-14 13:57:28.050678",
+ "modified_by": "Administrator",
+ "module": "Core",
+ "name": "Document Key",
```
Should be `Document Share Key`?

surajshetty3416 replied:
or "Document Print Key"... since it is currently used specifically for print-view?

--
Also, that dropdown of "Visibility" seems useless in our case. Do we need the ability to generate a key only for logged-in users (this was my initial thought for adding that dropdown)?

sagarvora replied:
- Visibility: not useful (confirmed by checking reference implementation)
- I think we need a way to distinguish this and existing Share feature. Maybe **Generate Link**? Although, **Share Link** has a nice ring to it. A totally different approach would be to show a link icon in below marked area:

![Screenshot-2022-01-20-203255](https://user-images.githubusercontent.com/16315650/150364737-589946f4-b4c6-406a-97ee-4b251f55293a.png)

This is somewhat similar to what Google does:

![image](https://user-images.githubusercontent.com/16315650/150365376-a18e1d5f-951f-4bf6-b408-399c22b5bcc8.png)

- I like **Document Share Key** better. Although it is technically a print view, print isn't the function.

netchampfaris reviewed (commented):
How is the "Private" share feature different from the "Share" feature we already have? If they are similar, should we merge these features for better UX?

surajshetty3416 committed:
fix: Document key expiry logic

surajshetty3416 committed:
feat: Templates for invalid and expired key

surajshetty3416 committed:
refactor: Permission validation for print view

- Show UI based on the status of the key/permission
- Respond with proper http_code according to the permission validation

sagarvora commented:
- I would want the older signature to be deprecable based on a transient setting in System Settings (whole thing could go in next major version - 15). This way, people can stop accepting older links once they are confident that they are no longer being used by email recipients.

- Should a new key be generated for same doc and expiry date? Maybe just returning existing key should do.
- Would be nice: viewing existing links with expiry in the form itself. This way a person can choose to send an existing link instead of generating a new one.

sagarvora commented:
>  Notify users to avoid links with old signatures and to get new ones?

The users of these links are Guests. I don't think we should concern them with such things.

surajshetty3416 committed:
fix: Rename "Document Key" to "Document Share Key"

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into fix-document-signature

surajshetty3416 committed:
fix: Allow document sharing who has share rights

surajshetty3416 committed:
fix: Move document_share_key_expiry to System Settings

- Having `document_share_key_expiry` on each doctype is overkill so kept it at System Settings

surajshetty3416 committed:
fix: Add setting "allow_older_web_view_links" in System Settings

surajshetty3416 committed:
fix: Re-use document key if key with same expiry exists

surajshetty3416 committed:
fix: Move share link functionality to sidebar (WIP)

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into fix-document-signature

surajshetty3416 committed:
fix: Share link style

surajshetty3416 committed:
feat: Option to Print or download PDF in print view

surajshetty3416 committed:
feat: Ability to create share link without expiry

surajshetty3416 committed:
wip

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into fix-document-signature

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into fix-document-signature

surajshetty3416 committed:
fix: Option set no_expiry

sagarvora commented on frappe/model/document.py:
```diff
@@ -1278,6 +1278,30 @@ def get_signature(self):
+		if no_expiry:
+			expires_on = None
+
+		if document_key_exist := frappe.db.exists(
```
```suggestion
		if existing_key := frappe.db.exists(
```



Good job on using the walrus operator 👍🏼

phot0n replied:
are we going to support >= py3.8 in v14 🤔 ? (walrus was introduced in py3.8)

sagarvora replied:
Yes:
https://frappeframework.com/docs/v13/user/en/installation#pre-requisites

sagarvora replied:
@surajshetty3416 I think you missed it, I had added a suggestion for variable naming.

sagarvora commented on frappe/model/document.py:
```diff
@@ -1278,6 +1278,30 @@ def get_signature(self):
 		return hashlib.sha224(get_datetime_str(self.creation).encode()).hexdigest()
 
+	@frappe.whitelist()
+	def get_new_document_share_key(self, expires_on=None, no_expiry=False):
```
Can return existing key as well.

```suggestion
	def get_document_share_key(self, expires_on=None, no_expiry=False):
```

sagarvora commented on frappe/public/js/frappe/form/form.js:
```diff
@@ -1149,6 +1149,47 @@ frappe.ui.form.Form = class FrappeForm {
+	}
+
+	get_share_link(key) {
+		return `${window.origin}/${this.doctype}/${this.docname}?key=${key}`;
```
Use `frappe.urllib.get_full_url` ?

sagarvora commented on frappe/public/js/frappe/form/form.js:
```diff
@@ -1149,6 +1149,47 @@ frappe.ui.form.Form = class FrappeForm {
 		});
 	}
 
+	share_doc_link() {
```
I think this is the older modal, remove.

sagarvora commented on frappe/public/js/frappe/form/sidebar/share.js:
```diff
@@ -187,4 +191,52 @@ frappe.ui.form.Share = class Share {
+	async share_modal() {
+		const default_expiry = frappe.boot.sysdefaults.document_share_key_expiry;
+		const share_modal = new frappe.ui.Dialog({
+			title: __("Share Link"),
```
I think we need to introduce the words **Public** and **Read Only** here to make it more explicit that this feature is to share a read-only print view to external users.

sagarvora commented on frappe/public/js/frappe/form/sidebar/share.js:
```diff
@@ -187,4 +191,52 @@ frappe.ui.form.Share = class Share {
+				fieldtype: "Date",
+				default: frappe.datetime.add_days(moment(), default_expiry),
+				min_date: new Date(),
+				description: __("The default expiration date is {0} days from today.", [default_expiry]),
```
This doesn't work if system setting isn't set:

![Screenshot-2022-03-10-100945](https://user-images.githubusercontent.com/16315650/157593341-111ec95e-208c-4d6e-b429-2dc6a1266946.png)

sagarvora commented on frappe/public/js/frappe/form/templates/form_sidebar.html:
```diff
@@ -87,6 +87,13 @@
 		<button class="share-doc-btn text-muted btn btn-default icon-btn">
 			<svg class="icon icon-sm"><use href="#icon-add"></use></svg>
 		</button>
+		<span class="avatar avatar-small share-link">
+			<div class="avatar-frame avatar-action">
+				<svg class="icon icon-sm">
+					<use href="#icon-link-url"></use>
+				</svg>
+			</div>
+		</span>
```
UI note:

I would like a bit more contrast and same size as the plus button.

![Screenshot-2022-03-10-102948](https://user-images.githubusercontent.com/16315650/157593415-499cec66-9fd0-4762-9f9b-18ac3b5ca7c8.png)

Perhaps this should also say something like **Get Shareable Link** on hover.

sagarvora commented on frappe/public/js/frappe/form/sidebar/share.js:
```diff
@@ -187,4 +191,52 @@ frappe.ui.form.Share = class Share {
+				label: __("No Expiry"),
+				fieldname: "no_expiry",
+				change: () => {
+					const no_expiry_warning = `<span style="color: var(--red-500);">
+						${__('It is not recommended to send links without expiry')}
```
Is this needed? Maybe make it a warning. Perhaps it would be better to say why or just warn that the link user will be able to see all changes to this document as long as it is not deleted.

sagarvora commented on frappe/public/js/frappe/form/toolbar.js:
```diff
@@ -263,6 +263,15 @@ frappe.ui.form.Toolbar = class Toolbar {
 			});
 		}
 
+		// share print view link
+		if (frappe.model.can_email(null, me.frm)
+			&& frappe.model.can_share(null, me.frm)
+			&& me.frm.doc.docstatus < 2) {
+			this.page.add_menu_item(__("Share Link", [me.frm.doctype]), function() {
+				me.frm.share_doc_link();
+			}, true);
+		}
```
remove?

sagarvora commented on frappe/core/doctype/system_settings/system_settings.json:
```diff
@@ -485,12 +487,25 @@
+   "default": "1",
+   "fieldname": "allow_older_web_view_links",
+   "fieldtype": "Check",
+   "label": "Allow Older Web View Links (Insecure)"
```
Perhaps this and `document_share_key_expiry` need a patch for setting default? The default didn't refect on my doc.

sagarvora commented on frappe/core/doctype/system_settings/system_settings.json:
```diff
@@ -485,12 +487,25 @@
    "options": "Sunday\nMonday\nTuesday\nWednesday\nThursday\nFriday\nSaturday"
+  },
+  {
+   "default": "90",
```
I think 30 would be a better default for most. But I could be biased. 😅

sagarvora commented on frappe/core/doctype/system_settings/system_settings.json:
```diff
@@ -485,12 +487,25 @@
+   "label": "Document Share Key Expiry (in Days)"
+  },
+  {
+   "default": "1",
```
Let's keep this 0 and make a patch. This way, the insecure setting doesn't get enabled for new users.

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into fix-document-signature

surajshetty3416 committed:
fix: Add patch to set document expiry

surajshetty3416 committed:
refactor: Update method name

get_new_document_share_key -> get_document_share_key

surajshetty3416 committed:
refactor: Remove unnecessary code

surajshetty3416 committed:
fix: Get shareable link style

sagarvora commented on frappe/patches/v14_0/set_document_expiry_default.py:
```diff
@@ -0,0 +1,5 @@
+import frappe
+
+def execute():
+	frappe.db.set_value("System Settings", "System Settings", "document_share_key_expiry", 30)
+	frappe.db.set_value("System Settings", "System Settings", "allow_older_web_view_links", 1)
```
Pass dict to reduce query 😅

surajshetty3416 replied:
Done... that usage does not feel intuitive though 😅

surajshetty3416 committed:
refactor: Set values in a single query

surajshetty3416 commented:
will re-open once this is ready
